src/plugin/safe-empty-array-where-in/safe-empty-array-where-in.transformer.ts

```diff
@@ -0,0 +1,64 @@
+import { BinaryOperationNode } from '../../operation-node/binary-operation-node.js'
+import { OperationNode } from '../../operation-node/operation-node.js'
+import { OperationNodeTransformer } from '../../operation-node/operation-node-transformer.js'
+import { PrimitiveValueListNode } from '../../operation-node/primitive-value-list-node.js'
+import { WhereNode } from '../../operation-node/where-node.js'
+
+const IN_OPERATORS = ['in', 'not in'] as const
+
+interface InOperatorNode {
+  kind: 'OperatorNode'
+  operator: (typeof IN_OPERATORS)[number]
+}
+
+export class WithSafeArrayWhereInTransformer extends OperationNodeTransformer {
+  #BinaryOperationNode(node: OperationNode): node is BinaryOperationNode {
+    return node.kind === 'BinaryOperationNode'
+  }
+
+  #isPrimitiveValueListNode(
+    node: OperationNode,
+  ): node is PrimitiveValueListNode {
+    return node.kind === 'PrimitiveValueListNode'
+  }
+
+  #IsInOperatorNode(node: OperationNode): node is InOperatorNode {
+    return (
+      (node as InOperatorNode).kind === 'OperatorNode' &&
+      IN_OPERATORS.includes((node as InOperatorNode).operator)
+    )
+  }
+
+  protected transformWhere(node: WhereNode): WhereNode {
+    const binaryOperatorNode = node.where
+
+    if (!this.#BinaryOperationNode(binaryOperatorNode)) {
+      return node
+    }
+
+    const rightOperand = binaryOperatorNode.rightOperand
+    const operator = binaryOperatorNode.operator
+
+    if (
+      this.#IsInOperatorNode(operator) &&
+      this.#isPrimitiveValueListNode(rightOperand)
+    ) {
+      const values = rightOperand.values
+
+      if (Array.isArray(values) && values.length === 0) {
+        return Object.freeze({
+          ...node,
+          where: {
+            ...binaryOperatorNode,
+            rightOperand: {
+              ...rightOperand,
+              values: [null],
```

Comment:
what about situations where the value in db is `null`?

Reply:
in (null) should always return nothing, its the equivalent of `1=0`

Reply:
OK let's make sure this is correct across all 4 dialects:

1. have a record in data that has a `null` in that column.
2. run query against db and expect results are empty.

Reply:
Alternatively, could just alter the tree to `1=0` or equivalent and call it a day.

Reply:
https://github.com/kysely-org/kysely/pull/925/commits/015ee37730b5c2a0757d9985aabf3f0ed3e9f299 added tests here. Let me know if it is sufficient

Reply:
@austinwoon I agree with @igalklebanov that you should switch to `1=0` and `1=1` rewrites for in/not-in clauses. Right now the transform does not seem to work correctly for the `not in` operator, since `select 1 where 1 not in (null);` does not return any rows. Following code seems to transform things correctly.

```ts
// NOTE: Translating the clause to `1 = 1` behaves semantically as truthy
if (operator.operator === 'not in') {
  return BinaryOperationNode.create(
    ValueNode.create(1),
    OperatorNode.create('='),
    ValueNode.create(1)
  );
}

// NOTE: Translating the clause to `1 = 0` behaves semantically as falsey
return BinaryOperationNode.create(
  ValueNode.create(1),
  OperatorNode.create('='),
  ValueNode.create(0)
);
```

We were testing this PR in our project, and noticed the issue, thanks for all your work!

Reply:
Ah I see, for us that not in behaviour was okay. So you're saying the above behaviour would result in `not in (null)` returning essentially the whole table?

Reply:
Sticking with original behaviour as a no-op is preferred for now. We can resolve this in another PR if you think otherwise!